File: packages/catalog-schemas/src/components/form-card.ts
import {z} from 'zod';
import {type CatalogComponentDefinition} from '../types';

export const formFieldTypeSchema = z.enum(['text', 'number', 'select', 'boolean', 'lookup']);

export const formFieldSchema = z
  .object({
    name: z.string().min(1).max(80),
    label: z.string().min(1).max(120),
    type: formFieldTypeSchema,
    required: z.boolean().optional(),
    options: z.array(z.string()).min(1).optional(),
    // lookup: имя справочника в контексте агента-владельца формы.
    referenceId: z.string().min(1).max(80).optional(),
    // lookup: порог начала поиска; default 3 применяет рендерер.
    minChars: z.number().int().min(1).max(10).optional(),
    placeholder: z.string().min(1).max(120).optional(),
    defaultValue: z
      .union([
        z.string(),
        z.number(),
        z.boolean(),
        // lookup: value уходит в submit, label показывается в поле.
        z.object({value: z.string(), label: z.string()}).strict(),
      ])
      .optional(),
  })
  .strict()
  .refine(field => field.type !== 'lookup' || typeof field.referenceId === 'string', {
    message: 'referenceId is required when type is "lookup"',
    path: ['referenceId'],
  });

export const formCardSubmitSchema = z
  .object({
    label: z.string().min(1).max(80),
    action: z.string().min(1).max(120),
  })
  .strict();

export const formCardPropsSchema = z
  .object({
    title: z.string().min(1).max(120),
    description: z.string().min(1).max(240).optional(),
    fields: z.array(formFieldSchema).min(1),
    submit: formCardSubmitSchema,
  })
  .strict();

export type FormFieldType = z.infer<typeof formFieldTypeSchema>;
export type FormField = z.infer<typeof formFieldSchema>;
export type FormCardSubmit = z.infer<typeof formCardSubmitSchema>;
export type FormCardProps = z.infer<typeof formCardPropsSchema>;

export const formCardDefinition: CatalogComponentDefinition<typeof formCardPropsSchema> = {
  name: 'FormCard',
  slug: 'form-card',
  description:
    'An interactive form card with typed fields for human-in-the-loop prompts. Use it when the agent must collect structured parameters (text, number, select, boolean, or lookup with reference-backed autocomplete) from the user and return them via a submit action.',
  schema: formCardPropsSchema,
};

File: packages/catalog-python/src/ai37_a2ui_catalog/models/form_card.py
from __future__ import annotations

from typing import Literal

from pydantic import Field, model_validator

from .shared import StrictModel

FormFieldType = Literal["text", "number", "select", "boolean", "lookup"]


class FormFieldLookupValue(StrictModel):
    value: str
    label: str


class FormField(StrictModel):
    name: str = Field(min_length=1, max_length=80)
    label: str = Field(min_length=1, max_length=120)
    type: FormFieldType
    required: bool = None
    options: list[str] = Field(default=None, min_length=1)
    referenceId: str = Field(default=None, min_length=1, max_length=80)
    minChars: int = Field(default=None, ge=1, le=10)
    placeholder: str = Field(default=None, min_length=1, max_length=120)
    defaultValue: str | int | float | bool | FormFieldLookupValue = None

    @model_validator(mode="after")
    def validate_lookup_reference(self) -> "FormField":
        if self.type == "lookup" and not isinstance(self.referenceId, str):
            raise ValueError('referenceId is required when type is "lookup"')
        return self


class FormCardSubmit(StrictModel):
    label: str = Field(min_length=1, max_length=80)
    action: str = Field(min_length=1, max_length=120)


class FormCardProps(StrictModel):
    title: str = Field(min_length=1, max_length=120)
    description: str = Field(default=None, min_length=1, max_length=240)
    fields: list[FormField] = Field(min_length=1)
    submit: FormCardSubmit
